[quant][fx] Support override observers and fake quantize module in backend_config_dict

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75260
* #75259
* #75258
* #75241
* #75202
* __->__ #75135
* #74882

Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

jerryzh168 committed:
[quant][fx] Support override observers and fake quantize module in backend_config_dict

Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35334279).

jerryzh168 commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -91,6 +93,15 @@ def _get_default_op_backend_config(op, dtype_configs):
+_HARDSIGMOID_MODULE_CONFIG = {
+    "pattern": torch.nn.Hardsigmoid,
+    "observation_type": ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT,
+    "overwrite_output_fake_quantizer": FixedQParamsFakeQuantize.with_args(observer=default_affine_fixed_qparams_observer),
```
hmmm, currently we override the observer for these ops, so we'd need to put this into config I think, but maybe we could do a refactor later to move these to default qconfig_dict setting, and put the requirement here, like "quant_min", "quant_max", "dtype" requirement etc. but to unblock this refactor I think we can land this first

vkuzo replied:
would it be worth considering not landing this then?  Seems like currently this happens in the workflow, which is not ideal.  However, putting it into the "wrong" config sounds even more confusing than leaving it hardcoded in the code.

jerryzh168 replied:
Maybe I can start with `_` for now for the configuration key? we still need to do this refactor and use backend_config_dict everywhere in the prepare first, fixing the existing issue will be a separate step I think

jerryzh168 replied:
to finish the frontend refactor, we don't have the option to just leave it hardcoded in the current state

vkuzo replied:
makes sense, how about moving this to qconfig_dict right away then instead of putting it here?  What's the eng cost on that?

jerryzh168 replied:
unfortunately this is not possible to currently, since the new path is not fully enabled yet in the code, current state is kind of intermediate state where we use the default quant patterns and the new quant pattern together see L138 and L144 in https://github.com/pytorch/pytorch/blob/0765a804911673fb2d9694a76ba0196ea0eddec4/torch/ao/quantization/fx/backend_config/utils.py#L133, and the backend_config_dict is actually None for fbgemm/qnnpack currently, which will be used in https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L181 to guard the configurations(dtype, quant_min, quant_max etc.) in QConfig, this path can only be enabled after we have everything fully migrated to backend_config_dict.

It's a little bit complicated, but the short answer is we can't do that right now

jerryzh168 replied:
yeah sure

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35334279).

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -303,6 +304,22 @@ def _get_conv_configs():
+    "observation_type": ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT,
+    # TODO: The following two keys are temporary, since we don't want to put observer in the configs
+    # we expect that it's provided by user
+    # What we want to put here is the requirement on observers, in this case dtype,
+    # quant_min, quant_max etc., but we need to first move all configs to
```
quant_min, quant_max are internal variables to observers.  I think restrictions in backend_config_dict should be on the public API only, such as scale and zero_point.

jerryzh168 replied:
OK, makes sense, yeah we can put scale/zero_point constraints here in the future

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support override observers and fake quantize module in backend_config_dict"


Summary:
Some operators have fixed quantization parameters, this PR adds the support to override the
qconfig in the backend_config_dict

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35334279](https://our.internmc.facebook.com/intern/diff/D35334279)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35334279).
